### PR TITLE
feat(megatron): add selective CPU offload as an alternative to torch_mem…

### PR DESCRIPTION
**Pull request by yeqinghe609**
…ory_saver

Colocate training offloads the train actor to CPU on every sleep/wake. The only path was torch_memory_saver (TMS) VMM pause()/resume(), which is semantically blind (backs up the whole runtime pool) and unsafe on backends without expandable_segments  This adds an explicit, application-level selective offload and lets it be selected per job.

## Behavior:
- Default stays TMS.
- New --selective-offload switches to selective offload: only live train state (param flat buffers + optimizer master/Adam state) is copied to CPU; grads and recomputable buffers are just released, then empty_cache() returns the freed pages to the driver.
### Use TMS (default) when
 torch_memory_saver hook LD_PRELOAD'ed (the launcher does this automatically for offload_train + megatron).
The allocator supports it (notably expandable_segments), so pause()/resume() can unmap/remap physical pages cleanly.
You want the simplest, well-tested path and don't mind that TMS backs up the entire runtime memory pool (it can't tell live optimizer state from dead activation/cache blocks).
### Use --selective-offload when
You want selective offload: only the live train state (param flat buffers + optimizer master/Adam state) is copied to CPU; gradients and recomputable buffers are just released and empty_cache() returns the freed pages to the driver. This avoids blindly copying the ~70% dead data that whole-pool TMS backup moves, cutting sleep+wake time significantly on large/MoE/long-sequence models.
You need a hardware-neutral path that relies only on plain PyTorch storage ops (no .so / VMM), consistent across NVIDIA/Others.

## Changes:
- weight_update/train_offload.py (new): MegatronTrainStateOffloader facade delegating to _TmsOffloadStrategy / _SelectiveOffloadStrategy behind an _OffloadStrategy Protocol. Owns TMS margin init and strategy selection (TMS when LD_PRELOAD'ed, else selective; --selective-offload forces selective).
- backends/megatron/actor.py: sleep()/wake_up()/update_weights() delegate to the offloader; drop the inline _torch_memory_saver_enabled block and the torch_memory_saver / nullcontext imports.
- weight_update/common.py: _maybe_get_cpu_backup() reads from the offloader's _relax_cpu_offload_data when a param's GPU storage was freed, and only takes the TMS get_cpu_backup() branch when the hook is actually LD_PRELOAD'ed.
- distributed/ray/actor_group.py: skip LD_PRELOAD + TMS_INIT env vars under --selective-offload so TMS is not the active mechanism.
- utils/arguments.py: add --selective-offload.
- scripts/training/text/run-qwen35-8xklx.sh: pass --selective-offload for klx.

No behavior change for existing TMS/CUDA users (default path untouched).

## Testing
Qwen3.6-35B

### performance
* wake_up_time: 10s->2s
* sleep_time: 20s->5s
31b9536d7d9d0c485d4f7270c68e7550

56faaa9b8094f2daff6b3153c3497b7c

###  Memory
3G vs 9G(tms offload vs selective offload)
tms offload

image

selective offload

image

- [ ] `pre-commit run --all-files` passes
- [ ] Tests pass (`pytest tests/`)
- [ ] New tests added (if applicable)
- [ ] Documentation updated (if applicable)

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ✅] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] CI/CD or build changes

### CODE REVIEW

**Comment by NINGBENZHE** [comment]
@codex review

File: relax/backends/megatron/weight_update/common.py [hunk 1]
```diff
@@ -111,9 +111,6 @@ def all_gather_param(args, name: str, param: torch.nn.Parameter) -> torch.Tensor
     if "linear_fc1.weight" in name and "vision_model" not in name:
         param_partitions = [p.chunk(2, dim=0) for p in param_partitions]
         param_partitions = [p[0] for p in param_partitions] + [p[1] for p in param_partitions]
```
[the remark this answers is not in the document]
Comment (yeqinghe609) [reply relax/backends/megatron/weight_update/common.py:114]: 这个部分不应该删掉，我改下

File: scripts/training/text/run-qwen35-9B-8xklx.sh [hunk 2]
```diff
@@ -265,14 +265,15 @@ ray job submit ${RAY_NO_WAIT:+--no-wait} --address="http://127.0.0.1:8265" \
...
-    "${CKPT_ARGS[@]}" \
-    "${ROLLOUT_ARGS[@]}" \
-    "${OPTIMIZER_ARGS[@]}" \
-    "${GRPO_ARGS[@]}" \
-    "${WANDB_ARGS[@]}" \
-    "${PERF_ARGS[@]}" \
-    "${EVAL_ARGS[@]}" \
-    "${SGLANG_ARGS[@]}" \
+   --use-health-check \
+   --selective-offload \
```
Comment (NINGBENZHE) [inline scripts/training/text/run-qwen35-9B-8xklx.sh:272]: 这个参数写错了吧，应该--manual-offload？

Comment (yeqinghe609) [reply scripts/training/text/run-qwen35-9B-8xklx.sh:272]: 已修复，全部更正为--selective-offload，本地验证okay

**Review by NINGBENZHE — DISMISSED** [review]
lgtm

File: relax/backends/megatron/weight_update/train_offload.py [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+
+        # 0. Drop MoE token-dispatcher caches (token_probs/dispatched_probs carry
+        # grad_fn and pin the MoE forward graph + its activations on GPU).
+        self._clear_moe_dispatcher_caches()
+
+        # 1. Model DDP flat buffers.
+        for buffer in self._iter_ddp_buffers():
+            # param_data: copy to CPU, then free GPU storage (must survive sleep).
+            if buffer.param_data is not None and buffer.param_data.storage().size() > 0:
+                buffer.param_data.cpu_data = buffer.param_data.data.cpu()
```
Comment (Yangruipis) [inline relax/backends/megatron/weight_update/train_offload.py]: can we use pin_memory for better performance here?

File: relax/backends/megatron/actor.py [hunk 4]
```diff
@@ -213,23 +212,6 @@ def _init(
...
-        self._torch_memory_saver_enabled = False
-        if args.offload_train:
-            x = max(int(args.train_memory_margin_bytes), 0)
-            try:
-                torch_memory_saver.memory_margin_bytes = x
-                self._torch_memory_saver_enabled = True
-                if x > 0:
-                    logger.info(f"Set torch_memory_saver.memory_margin_bytes to {x}")
-            except (RuntimeError, NotImplementedError) as e:
-                if "expandable_segments is not supported" in str(e) or "Only setter is supported" in str(e):
```
Comment (Yangruipis) [inline relax/backends/megatron/actor.py:225]: so expandable_segments is ok even when tms is enabled?

File: relax/backends/megatron/weight_update/train_offload.py (hunk 3 above)
Comment (yeqinghe609) [reply relax/backends/megatron/weight_update/train_offload.py]: yes, pinned memory helps here, and this is the highest-value spot for it
since `param_data` is a single large contiguous copy with a size that's fixed across
sleep/wake cycles.

I checked verl upstream (`verl/utils/megatron_utils.py`) and they've already moved to
exactly this: a persistent pinned `cpu_data` allocated **at most once per DDP buffer**,
reused on every offload. Worth noting *why* they don't just reallocate with
`.cpu().pin_memory()` each time — per their comment, Python evaluates the RHS before
the assignment, so the new pinned block is allocated while the old `cpu_data` is still
referenced, giving a transient 2x host-memory peak that OOMKilled pods against the
cgroup limit even though steady-state usage was 1x. So "allocate once and reuse" isn't
just a perf optimization, it's what makes pinning safe at scale.

Expected gain: pageable D2H stages through the driver's internal bounce buffer
(~8-12 GB/s); pinned can DMA directly (~20-25 GB/s on PCIe4). Both `offload()` and
`reload()` benefit from the same change.

Three things need to happen together, though — I'd rather not do a partial version:

1. **Drop the two `del`s in `reload()` (L275-276).** `del buffer.param_data_size` and
   `del buffer.param_data.cpu_data` are fundamentally incompatible with reuse. Keeping
   them means `existing is None` is always true, so nothing is ever reused and the 2x
   peak comes right back. Worse: if `param_data_size` ends up assigned only on the
   first-allocation path, the `hasattr(buffer, "param_data_size")` guard at L267 fails
   on the second wake, `param_data` storage stays at 0, and we lose weights **silently**
   with no error. verl keeps both attributes alive across cycles and asserts
   `param_data_size == cpu_data.storage().size()` after every `resize_(0)`.

2. **`pin_memory` must go through `device_utils.use_pinned_host_memory()`**, not a
   hardcoded `True`. Pinned + non-blocking copies segfault on KLX — that's why the
   helper exists (`relax/utils/device.py:574`). See
   `relax/utils/training/tensor_backper.py:65` for the established pattern in this repo.
   verl doesn't have to care about this; we do.

3. **`param._relax_cpu_offload_data` (L216) needs a comment.** These are slice views
   into `cpu_data`, which we have and verl doesn't. Once `cpu_data` is persistent, those
   views alias memory that the *next* offload overwrites. It's safe today because
   `reload()` deletes them at L272-274, but that coupling is invisible — anyone moving
   the deletion later would start reading stale weights.

One tradeoff to flag explicitly: today `cpu_data` is only resident during sleep and
freed on wake. With a persistent pinned buffer it's resident for the whole run and
non-swappable, roughly one full copy of the model weights locked per node. That's a
real cost, not a free win.

On sequencing: this PR is the functional introduction of manual offload, so I'd prefer
to land the pinned-memory change as a follow-up, with timing instrumentation in
`offload()`/`reload()` first to confirm `param_data` is actually the dominant cost
rather than the optimizer-state copies. Happy to do it here instead if you'd rather not
split it — just don't want to widen this PR's scope on an unmeasured assumption.

File: relax/backends/megatron/actor.py (hunk 4 above)
Comment (yeqinghe609) [reply relax/backends/megatron/actor.py:225]: No，it cause runtime error. I will add an explicit raise and fall back to selective offload.

Comment (yeqinghe609) [reply relax/backends/megatron/actor.py:225]: add capture exception in arguments.py Line2878

**Review by NINGBENZHE — APPROVED** [review]
lgtm
